text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,391 @@
+    ```
+
+### Usage example of the Transactional Key-Value API
+
+```rust
+extern crate futures;
+extern crate tikv_client;
+
+use std::ops::RangeBounds;
+
+use futures::{future, Future, Stream};
+use tikv_client::transaction::{Client, IsolationLevel};
+use tikv_client::*;
+
+fn puts(client: &Client, pairs: impl IntoIterator<Item = impl Into<KvPair>>) {
+    let mut txn = client.begin();
+    let _: Vec<()> = future::join_all(pairs.into_iter().map(Into::into).map(|p| txn.set(p)))
+        .wait()
+        .expect("Could not set key value pairs");
+    txn.commit().wait().expect("Could not commit transaction");
+}
+
+fn get(client: &Client, key: &Key) -> Value {
+    let txn = client.begin();
+    txn.get(key).wait().expect("Could not get value")
+}
+
+fn scan(client: &Client, range: impl RangeBounds<Key>, mut limit: usize) {
+    client
+        .begin()
+        .scan(range)
+        .take_while(move |_| {
+            Ok(if limit == 0 {
+                false
+            } else {
+                limit -= 1;
+                true
+            })
+        }).for_each(|pair| {
+            println!("{:?}", pair);
+            Ok(())
+        }).wait()
+        .expect("Could not scan keys");
+}
+
+fn dels(client: &Client, keys: impl IntoIterator<Item = Key>) {
+    let mut txn = client.begin();
+    txn.set_isolation_level(IsolationLevel::ReadCommitted);
+    let _: Vec<()> = keys
+        .into_iter()
+        .map(|p| {
+            txn.delete(p).wait().expect("Could not delete key");
+        }).collect();
+    txn.commit().wait().expect("Could not commit transaction");
+}
+
+fn main() {
+    let config = Config::new(vec!["127.0.0.1:3379"]);
+    let txn = Client::new(&config)
+        .wait()
+        .expect("Could not connect to tikv");
+
+    // set
+    let key1: Key = b"key1".to_vec().into();
+    let value1: Value = b"value1".to_vec().into();
+    let key2: Key = b"key2".to_vec().into();
+    let value2: Value = b"value2".to_vec().into();
+    puts(&txn, vec![(key1, value1), (key2, value2)]);
+
+    // get
+    let key1: Key = b"key1".to_vec().into();
+    let value1 = get(&txn, &key1);
+    println!("{:?}", (key1, value1));
+
+    // scan
+    let key1: Key = b"key1".to_vec().into();
+    scan(&txn, key1.., 10);
+
+    // delete
+    let key1: Key = b"key1".to_vec().into();
+    let key2: Key = b"key2".to_vec().into();
+    dels(&txn, vec![key1, key2]);
+}
+```
+
+The result is like:
+
+```bash
+(Key([107, 101, 121, 49]), Value([118, 97, 108, 117, 101, 49]))
+(Key([107, 101, 121, 49]), Value([118, 97, 108, 117, 101, 49]))
+(Key([107, 101, 121, 49]), Value([118, 97, 108, 117, 101, 49]))
+```
+
+## Programming model
+
+The client instance is thread safe and all the interfaces return futures so users can use the client in either asynchronous or synchronous way. A dedicated event loop thread is created at a per client instance basis to drive the reactor to make progress.
```

Comment:
> The client instance is thread safe ...

Being thread safe is kind of expensive. If the client has some mutable state we may end up with a `Mutex` inside and every APIs require the `Mutex`. I suggest the client should impls `Send + Clone`(another concurrency schema but more flexible). It allows private mutable state, eg, we change client config on the fly.

>  dedicated event loop thread is created at a per client instance basis to drive the reactor to make progress.

I suggest we should not specify thread model, because it is implementation detail. We should leave some space to make sure we dont back ourselves into a corner.

------------------

Based on the model above, here is the code I (as a user) can imagine:

```rust
let client = Arc::new(TikvClient::new());
for _ in 0..4 { // My app has 4 workers.
  let cli = client.clone();
  thread::spawn(move || {
    loop {
      let resp = cli.request().wait();
      handle(resp);
    }
  }).unwrap();
}
```

Users may always wrap the client in an `Arc` to avoid redundant event loop threads.

Reply:
The reason that there is one dedicated thread is because the tso interface of PD is a bidirectional stream. Using a dedicated thread is a way I figured out to keep the order of responses the same as requests yet can still do batching. Not sure if there are other ways to implement that.

Agree with Send + Clone part. It actually has methods that takes `&mut self` so I think it is not correct to say client is thread safe. Let me fix it.

Reply:
How about we remove this section for now and let it to be determined? We might figure out a better way later on.

Reply:
Yes, I agree.